File: app/src/main/java/catchla/yep/view/holder/InstagramMediaViewHolder.java
package catchla.yep.view.holder;

import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.util.List;

import catchla.yep.R;
import catchla.yep.model.InstagramImage;
import catchla.yep.model.InstagramMedia;

/**
 * Created by mariotaku on 15/6/3.
 */
public class InstagramMediaViewHolder extends RecyclerView.ViewHolder implements Callback {
    private final ImageView imageView;
    private final View imageProgress;

    public InstagramMediaViewHolder(final View itemView) {
        super(itemView);
        imageView = (ImageView) itemView.findViewById(R.id.image_view);
        imageProgress = itemView.findViewById(R.id.image_progress);
    }

    public void displayShot(InstagramMedia shot) {
        final InstagramImage image = getBestImage(shot.getImages());
        if (image != null) {
            imageProgress.setVisibility(View.VISIBLE);
            Picasso.with(itemView.getContext()).load(image.getUrl()).into(imageView, this);
        } else {
            imageProgress.setVisibility(View.GONE);
            Picasso.with(itemView.getContext()).cancelRequest(imageView);
        }
    }

    private InstagramImage getBestImage(final List<InstagramImage> images) {
        if (images == null) return null;
        for (InstagramImage image : images) {
            if ("low_resolution".equals(image.getResolution())) return image;
        }
        return null;
    }

    @Override
    public void onSuccess() {
        imageProgress.setVisibility(View.GONE);
    }

    @Override
    public void onError() {
        imageProgress.setVisibility(View.GONE);
    }
}

File: src/main/java/catchla/yep/view/holder/InstagramMediaViewHolder.kt
package catchla.yep.view.holder

import android.graphics.Bitmap
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.ImageView
import catchla.yep.R
import catchla.yep.adapter.InstagramMediaAdapter
import catchla.yep.model.InstagramImage
import catchla.yep.model.InstagramMedia
import com.nostra13.universalimageloader.core.assist.FailReason
import com.nostra13.universalimageloader.core.listener.ImageLoadingListener

/**
 * Created by mariotaku on 15/6/3.
 */
class InstagramMediaViewHolder(
        itemView: View,
        private val adapter: InstagramMediaAdapter
) : RecyclerView.ViewHolder(itemView), ImageLoadingListener {
    private val imageView: ImageView
    private val imageProgress: View

    init {
        imageView = itemView.findViewById(R.id.image_view) as ImageView
        imageProgress = itemView.findViewById(R.id.image_progress)
    }

    fun displayShot(shot: InstagramMedia) {
        val image = getBestImage(shot.images)
        val imageLoader = adapter.imageLoader
        if (image != null) {
            imageLoader.displayImage(image.url, imageView, this, null)
        } else {
            imageLoader.cancelDisplayTask(imageView)
        }
    }


    override fun onLoadingStarted(imageUri: String, view: View) {
        imageProgress.visibility = View.VISIBLE
    }

    override fun onLoadingFailed(imageUri: String, view: View, failReason: FailReason) {
        imageProgress.visibility = View.GONE
    }

    override fun onLoadingComplete(imageUri: String, view: View, loadedImage: Bitmap) {
        imageProgress.visibility = View.GONE
    }

    override fun onLoadingCancelled(imageUri: String, view: View) {
        imageProgress.visibility = View.GONE
    }

    companion object {

        fun getBestImage(images: List<InstagramImage>?): InstagramImage? {
            if (images == null) return null
            for (image in images) {
                if ("low_resolution" == image.resolution) return image
            }
            return null
        }
    }
}
